src/Carousel.js

```diff
@@ -100,441 +124,430 @@ const propTypes = {
    * Set to null to deactivate.
    */
   nextLabel: PropTypes.string,
-
-  /**
-   * Whether the carousel should support left/right swipe interactions on touchscreen devices.
-   */
-  touch: PropTypes.bool,
 };
 
 const defaultProps = {
   slide: true,
   fade: false,
+  controls: true,
+  indicators: true,
+
+  defaultActiveIndex: 0,
   interval: 5000,
   keyboard: true,
-  pauseOnHover: true,
+  pause: 'hover',
   wrap: true,
-  indicators: true,
-  controls: true,
-  activeIndex: 0,
+  touch: true,
 
   prevIcon: <span aria-hidden="true" className="carousel-control-prev-icon" />,
   prevLabel: 'Previous',
 
   nextIcon: <span aria-hidden="true" className="carousel-control-next-icon" />,
   nextLabel: 'Next',
-  touch: true,
 };
 
-class Carousel extends React.Component {
-  state = {
-    prevClasses: '',
-    currentClasses: 'active',
-    touchStartX: 0,
-  };
-
-  isUnmounted = false;
+function isVisible(element) {
+  if (
+    !element ||
+    !element.style ||
+    !element.parentNode ||
+    !element.parentNode.style
+  ) {
+    return false;
+  }
 
-  carousel = React.createRef();
+  const elementStyle = getComputedStyle(element);
 
-  componentDidMount() {
-    this.cycle();
-  }
+  return (
+    elementStyle.display !== 'none' &&
+    elementStyle.visibility !== 'hidden' &&
+    getComputedStyle(element.parentNode).display !== 'none'
+  );
+}
 
-  static getDerivedStateFromProps(
-    nextProps,
-    { activeIndex: previousActiveIndex },
-  ) {
-    if (nextProps.activeIndex !== previousActiveIndex) {
-      const lastPossibleIndex = countChildren(nextProps.children) - 1;
-
-      const nextIndex = Math.max(
-        0,
-        Math.min(nextProps.activeIndex, lastPossibleIndex),
-      );
-
-      let direction;
-      if (
-        (nextIndex === 0 && previousActiveIndex >= lastPossibleIndex) ||
-        previousActiveIndex <= nextIndex
-      ) {
-        direction = 'next';
-      } else {
-        direction = 'prev';
-      }
+const Carousel = React.forwardRef((uncontrolledProps, ref) => {
+  const {
+    // Need to define the default "as" during prop destructuring to be compatible with styled-components github.com/react-bootstrap/react-bootstrap/issues/3595
+    as: Component = 'div',
+    bsPrefix,
+    slide,
+    fade,
+    controls,
+    indicators,
+    activeIndex,
+    onSelect,
+    onSlide,
+    onSlid,
+    interval,
+    keyboard,
+    onKeyDown,
+    pause,
+    onMouseOver,
+    onMouseOut,
+    wrap,
+    touch,
+    onTouchStart,
+    onTouchMove,
+    onTouchEnd,
+    prevIcon,
+    prevLabel,
+    nextIcon,
+    nextLabel,
+    className,
+    children,
+    ...props
+  } = useUncontrolled(uncontrolledProps, {
+    activeIndex: 'onSelect',
+  });
+
+  const prefix = useBootstrapPrefix(bsPrefix, 'carousel');
+
+  const nextDirectionRef = useRef(null);
+  const [direction, setDirection] = useState(null);
```

Comment:
`orderClassName` has an initial value of  `carousel-item-null`, probably making it `next` makes sense here:
```
const [direction, setDirection] = useState('next');
```